### PR TITLE
Tom/project refactor

### PR DESCRIPTION
**Pull request by tomyang729**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,9 @@
 /node_modules
 /bin
 
+# git-crypt
+tegGenKey
```
Comment (DariaSova) [inline .gitignore]: the key file should not be in the directory, so i don't think this is necessary. Usually you only want to share this with the developers and they store it where they want, a lot of people have a separate folder with keys for different projects

File: clarifai.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ func NewClarifaiClient() *ClarifaiClient {
 }
 
 // Check whether the client still has access rights to Clarifai
-func (client *ClarifaiClient) isAccess() bool {
+func isCLFAccessible(client *ClarifaiClient) bool {
```
Comment (eladmica) [inline clarifai.go]: umm why did you change that from a method of ClarifaiClient to a function? I think it makes much more sense to do clarifaiClient.isAccessible() rather than  `isCLFAccessible(clarifaiClient)`

File: clarifai.go [hunk 3]
```diff
@@ -175,9 +175,9 @@ func getRequestBody(input string) (*bytes.Buffer, error) {
 }
 
 // Get image tag array from Clarifai's API response
-func getImageTagsArray(jsonStr string) ([]ClarifyTag, error) {
+func getImageTagsArray(jsonStr string) ([]ImageTag, error) {
```
Comment (eladmica) [inline clarifai.go]: In Go having a method/variable start with a capital letter means this method/variable is exported - i.e. accessible from outside the package. If we were to move the clarifai stuff to it's own folder, we wouldn't be able to call `getImageTagsFromCLF`. Internal helper functions should start with a lower case (as you don't actually want them to be called outside the package). Also, I really think this should be a method called on the `ClarifaiClient` struct

File: handlers.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	var request struct {
+		Image string `json:"image"`
+	}
+	err = json.Unmarshal(body, &request)
+	if err != nil {
+		fmt.Printf("Error--parse json: %s", err)
+	}
+
+	// TODO: somehow refresh token before it expires, without waiting for a request on an expired token
+	if !isCLFAccessible(CLFclient) {
```
Comment (eladmica) [inline handlers.go]: Yes so going back to the method vs. function point, I think `!CLFclient.isAccessible()` is less repetitive and makes more sense than `!isCLFAccessible(CLFclient)`

File: main.go [hunk 5]
```diff
@@ -4,27 +4,38 @@ import (
...
+	customTemplate := template.Must(template.New("main").ParseGlob("resources/templates/base/*.tmpl"))
+	customTemplate.ParseGlob("resources/templates/*.tmpl")
+	router.SetHTMLTemplate(customTemplate)
+	router.Static("/static", "resources/static")
 
-	server.Configure()
-	app.Run(":5050")
+	// Define routes **currently we only have two; it's not necessary to separate it to another file yet
+	router.GET("/", showHomePage)
+	router.POST("/getTags", fetchTags)
```
Comment (eladmica) [inline main.go]: Not a big fan of having this much logic in the main() function (think of how long this function would become if we have more routes, environment variables, etc.) But I'm totally OK with it if you guys prefer it this way

File: .gitignore (hunk 1 above)
Comment (tomyang729) [inline .gitignore]: I see. But I think it's still safer to ignore it just in case someone like me carelessly put it in the project directory

File: clarifai.go (hunk 2 above)
Comment (tomyang729) [inline clarifai.go]: For sure, both work. Previously, clairfai.go is like a class definition, all functions related with clarifai are defined as ClarifaiClient methods. And we can instantiate the ClarifaiClient struct (class) (it would be a singleton in our case) in the main.go, and we can access all the methods through this ClairfaiClient. It's nice and clear. An alternative approach is moving those clarifai methods out of box as normal functions (which can be accessible anywhere in the same package), and delare the ClarifaiClient struct in the main.go as a global variable. The reasons that I think the later is better are: 
1. golang is not primarily OO language, and I didn't see benefits of using object-oriented feature in Go to make the functions as methods here. ClarifaiClient is just a struct that holds the info of token and expiration_time, we just need one object of it instead of many different instances of it. 
2. Easy to work with other frameworks like Gin. For example, if we want to register a handler function to a specific route in Gin (the handler function and the ClairfaiClient are defined in different files) then we have to pass the ClarifaiClient to the handler function, which is difficult because typical ginHandler only takes gin.Context as argument

Comment (tomyang729) [inline clarifai.go]: But you are right. Here it doesn't make sense to pass `clarifaiClient` as an argument. I should just use the global clarifaiClient inside the function so it takes no arguments

File: clarifai.go (hunk 3 above)
Comment (tomyang729) [inline clarifai.go]: Somehow answered this from the above comment.
I think this rule doesn't apply for functions. All functions should be globally accessible across the same package

File: handlers.go (hunk 4 above)
Comment (tomyang729) [inline handlers.go]: ^^

File: main.go (hunk 5 above)
Comment (tomyang729) [inline main.go]: I'm not sure about the best practice here. But yeah, we can change back to having a server.go if that makes more sense

File: clarifai.go (hunk 2 above)
Comment (eladmica) [inline clarifai.go]: 1. Appending methods to a struct is not necessarily an OO feature and is actually very fundamental Go. I am almost certain you'll see it in any Go project you'll take a look at. In fact, when we're doing `http.Get` in our code we're actually calling the Get method on the default http client struct (see here: https://golang.org/pkg/net/http/#Client.Get). 

2. It has absolutely no effect on whether we're using frameworks or not, as it is not a parameter. You can easily register handler functions that are appended to a struct. Example: in Gin: `func (s *MyStruct) someFunc(c *gin.Context)` is a valid handler, which you can register like that: `app.GET("/endpoint", myStruct.someFunc)`. In standard net/http: `func (s *Struct) someFunc(w http.ResponseWriter, r *http.Request)` is a valid handler, which you can register like that: `http.HandleFunc("/endpoint", myStruct.someFunc)`

I was looking at some API clients when I first learned Go and they pretty much use the same approach. For example: https://github.com/fsouza/go-dockerclient/blob/master/client.go


This is my opinion though. Since no one else gave an opinion, it's up to you to choose what to do. I'm 100% OK with it if you choose to leave it as it, I just prefer the other approach

File: main.go (hunk 5 above)
Comment (eladmica) [inline main.go]: no no I like how you got rid of `server.go`, it was probably an overkill for our use case. I was thinking more in the line of having helper functions (inside `main.go`) that will do some of those stuff

File: clarifai.go (hunk 3 above)
Comment (eladmica) [inline clarifai.go]: Yes everything is always accessible in the same package. But only functions, struct fields, methods, etc. that start with a capital letter will be accessible outside the package. I guess this will not be a problem for us as we'll likely only have 1 package and our project would be never used as a dependency to something else

File: px.go [hunk 6]
```diff
@@ -7,10 +7,13 @@ import (
...
+	"net/url"
+	"strconv"
+)
 
-	"strings"
-
-	"github.com/joho/godotenv"
+const (
+	pxApi = "https://api.500px.com/v1/photos/search/?"
+	rpp = 50
```
Comment (eladmica) [inline px.go]: BTW what's `rpp`?

File: clarifai.go (hunk 2 above)
Comment (tomyang729) [inline clarifai.go]: Sure, I'm good with following your approach. In this case, there is indeed not much difference 😄 
And I agree that for big projects with many packages or APIs that will be exposed externally, encapsulate the functions as methods under structs is easier for the consumers to use. But for simple web APIs or things that don't have direct interaction with client side, is it still a good practice? I'm not trying to figure out right and wrong. Just want to have a better understanding over this interesting question:
"When should we use methods and when to use functions in Golang?" Technically, making a function as a pointer method under a struct has exactly the same effects as passing the struct pointer to a function. But in which cases do we prefer one over another? I've searched online but haven't found good answers yet. One answer mentioned that they are different in a sense that functions with struct pointers as arguments are more generic as we can pass any structs sharing the same interface to it, while method is only belong to one struct. 
Any insights on this? If not, maybe we can ask in the go channel?

File: clarifai.go (hunk 3 above)
Comment (tomyang729) [inline clarifai.go]: Oh good to know. Thanks!

File: main.go (hunk 5 above)
Comment (tomyang729) [inline main.go]: I see. I'll make the change

File: px.go (hunk 6 above)
Comment (tomyang729) [inline px.go]: Correct me if I'm wrong @qtoyoda I think it's number of photos we get from 500px per tag per request

File: clarifai.go (hunk 2 above)
Comment (eladmica) [inline clarifai.go]: hmm yeah I guess I didn't think about it that much. You're right, it doesn't make much of a difference. Appending these methods to a struct makes a lot more sense to me, but maybe that's because I mostly did OO up until now. I guess it comes down to personal preference.

**Comment by eladmica** [comment]
feel free to merge 👍
